### PR TITLE
659.1 sync api error improvements

### LINKED ISSUE #659: Compilation for sync settings
This branch does not compile: https://github.com/openmsupply/open-msupply/pulls

Need to propagate mutation/query changes to syncSettings.

Also as part of this would be great to do error mapping in structured way.

Ended up with more changes then I wanted to add to one PR, so broken down into:

- [ ] https://github.com/openmsupply/open-msupply/pull/671
   * Want to include url and flatten down MappedError, move variants of SyncErrorV5 to the same level as MappedError.  
   * Reformat thiserror derives slightly, they don't need to always provide 'source' as this can be traversed (like what anyhow::Error does)
   * Add specific error for SyncError rather then anyhow::Error (easier to map underlying SynApiError and better formatting for thiserror
- [ ] https://github.com/openmsupply/open-msupply/pull/672
- [ ] https://github.com/openmsupply/open-msupply/pull/673
- [ ] https://github.com/openmsupply/open-msupply/pull/677
- [ ] https://github.com/openmsupply/open-msupply/pull/679

### PR DESCRIPTION
**Pull request by andreievg**
closes #659 (this is to link the issue, there is a series of PR, as described in the issue)

[HEAD PR LINK](https://github.com/openmsupply/open-msupply/pull/679)

### Changes

Tidy up thiserror derives/annotation
* Previously was constantly stringifying source error, this is not needed since we can print source errors automatically (anyhow actually does this, but also added generic error formatter), see comment in self review
* Flatten Mapped error straight into sync api error
* Wanted url in SyncApiError, thus made it a wrapper around SyncApiErrorVariant
* Adde SyncErrorCodeV5 to map to expected central server error (this can be tested in the future

Used `from` and `?` to make sync code more readable (this is a pattern we've gone away from, instead using `map_err`, I would like to reinstate it, might need further discussions, but I think it's pretty clear, and in VS code you can just hover over methods to see what they return and deduce what `?` does)

Added generic error formatter (see doc tests)

### CODE REVIEW

File: server/service/src/sync/api/common_records.rs [hunk 1]
```diff
@@ -61,7 +61,7 @@ impl SyncActionV5 {
 }
 
 #[derive(Error, Debug)]
-#[error("Failed to parse V5 remote record into sync buffer row: {source:?} {record:?}")]
+#[error("Failed to parse V5 remote record into sync buffer row, record: '{record:?}'")]
```
Comment (andreievg) [inline server/service/src/sync/api/common_records.rs:64]: This is an example where source doesn't need to be specified, it will still be printed by format_error, if error enum variant or error struct has 'source' field or annotation of`#[source]` or `#[from]`

File: server/service/src/sync/api/core.rs [hunk 2]
```diff
@@ -40,15 +39,28 @@ fn generate_headers(hardware_id: &str) -> HeaderMap {
     headers
 }
 
+#[derive(Error, Debug)]
+pub enum SyncApiV5CreatingError {
```
Comment (andreievg) [inline server/service/src/sync/api/core.rs:43]: Added this so that we can return invalid url in API layer, i've noticed quite a few times when url was wrong that we would just get generic error from API and then in turn generic error in front end, now it would say invalid url

Comment (andreievg) [reply server/service/src/sync/api/core.rs:43]: In head PR you can see this in: link here

File: server/service/src/sync/api/core.rs [hunk 3]
```diff
@@ -40,15 +39,28 @@ fn generate_headers(hardware_id: &str) -> HeaderMap {
...
+    #[error("Error while creating SyncApiV5 instance")]
+    Other(#[source] anyhow::Error),
+}
+
 impl SyncApiV5 {
-    pub(crate) fn new(
+    pub fn new(
         settings: &SyncSettings,
         service_provider: &ServiceProvider,
-    ) -> anyhow::Result<Self> {
```
Comment (andreievg) [inline server/service/src/sync/api/core.rs:47]: Changes here are just for the new error result

File: server/service/src/sync/api/core.rs [hunk 4]
```diff
@@ -75,7 +87,10 @@ impl SyncApiV5 {
     where
         T: Serialize + ?Sized,
     {
-        let url = self.server_url.join(route).context("Failed to parse url")?;
+        let url = self
+            .server_url
+            .join(route)
+            .map_err(|error| self.api_error(route, error.into()))?;
```
Comment (andreievg) [inline server/service/src/sync/api/core.rs:93]: self.api_error is a helper for SyncApiError with url from self

File: server/service/src/sync/api/error.rs [hunk 5]
```diff
@@ -1,58 +1,128 @@
 use super::*;
```
Comment (andreievg) [inline server/service/src/sync/api/error.rs:1]: In this file change SyncApiError to have url, then added SyncApiErroVariant (as previous SyncApiError), and flattened mapped error into SyncApiVariant.
Also added concrete error code mapping form central server

File: server/service/src/sync/api/error.rs [hunk 6]
```diff
@@ -1,58 +1,128 @@
...
-    pub(crate) async fn from_response_and_status(
-        status: StatusCode,
-        response: Response,
-    ) -> SyncApiError {
-        let error = match to_json::<SyncErrorV5>(response).await {
-            Ok(source) => return SyncApiError::MappedError { source, status },
+#[derive(Error, Debug, Serialize, Deserialize)]
+#[error("code: '{code:?}' message: '{message}' data: '{data:?}")]
+pub struct ParsedError {
+    #[serde(serialize_with = "sync_error_code_v5_se")]
```
Comment (andreievg) [inline server/service/src/sync/api/error.rs:45]: In order to save any other sync api error codes in the `Other` variant of SyncErroCodeV5, used these two helpers

File: server/service/src/sync/synchroniser.rs [hunk 7]
```diff
@@ -26,6 +30,26 @@ pub struct Synchroniser {
     remote: RemoteDataSynchroniser,
 }
 
+#[derive(Error, Debug)]
+pub(crate) enum SyncError {
+    #[error("Database error while syncing")]
+    DatabaseError(#[from] RepositoryError),
```
Comment (andreievg) [inline server/service/src/sync/synchroniser.rs:36]: Added sync error vs anyhow return from synchroniser:
* Easier to map to sync api error (although can downcast but that's messy)
* Better use of thiserror annotation, previously in say RemotePushError, we would say "Sync api error in remote push" and we would add "remote push" to every variant. I think it's much cleaner to just add this once in the parent variant of underlying error

File: server/service/src/sync/central_data_synchroniser.rs [hunk 8]
```diff
@@ -10,15 +10,15 @@ use repository::{
 use thiserror::Error;
 
 #[derive(Error, Debug)]
-pub(crate) enum CentralSyncError {
```
Comment (andreievg) [inline server/service/src/sync/central_data_synchroniser.rs:13]: Also renamed to CentralPullError, to be consistent with others

File: server/service/src/sync/api/error.rs [hunk 9]
```diff
@@ -1,58 +1,128 @@
...
-        source: SyncErrorV5,
-        status: StatusCode,
-    },
-    #[error("Error connecting to server: ({url})")]
-    ConnectionError { source: reqwest::Error, url: Url },
-    #[error("{0}")]
-    ResponseParsingError(ParsingResponseError),
-    #[error("{0}")]
-    Other(#[from] anyhow::Error),
+#[error("Sync api error, url: '{url}', route: '{route}'")]
```
Comment (mark-prins) [inline server/service/src/sync/api/error.rs:11]: I'm not seeing this formatting when testing. Perhaps doing it wrong?

The error provides plenty of information - it can be pretty dense to read, with the nested errors though. Misspelling a domain gives:

```
2022-10-03 13:42:30.215860 INFO service::sync::site_info - Requesting site info
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: RequestSiteInfoError(SyncApiError { source: ConnectionError(reqwest::Error { kind: Request, url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhostie")), port: Some(2048), path: "/sync/v5/site", query: None, fragment: None }, source: hyper::Error(Connect, ConnectError("dns error", Custom { kind: Uncategorized, error: "failed to lookup address information: nodename nor servname provided, or not known" })) }), url: Url { scheme: "http", cannot_be_a_base: false, username: "", password: None, host: Some(Domain("localhostie")), port: Some(2048), path: "/", query: None, fragment: None }, route: "/sync/v5/site" })', /Users/m2/GitHub/open-msupply/server/server/src/lib.rs:129:22
```

I was wondering if the Url could be formatted with `Into<string>` instead of the whole object?

File: server/service/src/sync/site_info.rs [hunk 10]
```diff
@@ -11,17 +11,19 @@ use crate::{
...
+use super::api::{SyncApiError, SyncApiV5CreatingError};
 
 #[derive(Error, Debug)]
 pub enum RequestAndSetSiteInfoError {
-    #[error("Api error while requesting site info: {0:?}")]
-    RequestSiteInfoError(SyncApiError),
-    #[error("Database error whie requistin site info: {0:?}")]
+    #[error("Api error while requesting site info")]
+    RequestSiteInfoError(#[source] SyncApiError),
+    #[error("Database error whie requistin site info")]
```
Comment (mark-prins) [inline server/service/src/sync/site_info.rs:20]: :nit
```suggestion
    #[error("Database error while requesting site info")]
```

File: server/service/src/sync/api/error.rs [hunk 11]
```diff
@@ -84,20 +163,23 @@ mod test {
...
+                source: SyncApiErrorVariant::AsText {
+                    status: StatusCode::SERVICE_UNAVAILABLE,
+                    ..
+                },
+                ..
+            }
         );
 
-        // Service Unavailableg
+        // Service Unavailable`
```
Comment (mark-prins) [inline server/service/src/sync/api/error.rs:182]:
```suggestion
        // Service Unavailable
```

File: server/service/src/sync/remote_data_synchroniser.rs [hunk 12]
```diff
@@ -145,45 +127,35 @@ impl RemoteDataSynchroniser {
...
             if let Some(last_pushed_cursor_id) = last_pushed_cursor {
-                update_push_cursor(connection, last_pushed_cursor_id as u64 + 1)
-                    .map_err(Error::DatabaseError)?;
+                update_push_cursor(connection, last_pushed_cursor_id as u64 + 1)?;
             };
 
             match (response.integration_started, change_logs_total) {
                 (true, 0) => break,
-                (false, 0) => return Err(Error::IntegrationNotStarter),
+                (false, 0) => return Err(RemotePushError::IntegrationNotStarter),
```
Comment (mark-prins) [inline server/service/src/sync/remote_data_synchroniser.rs:158]: Should this be ??
```suggestion
                (false, 0) => return Err(RemotePushError::IntegrationNotStarted),
```

File: server/service/src/sync/test/integration/errors.rs [hunk 13]
```diff
@@ -51,6 +52,7 @@ mod tests {
     }
     #[actix_rt::test]
     async fn integration_sync_parsed_error() {
+        // util::init_logger(util::LogLevel::Warn);
```
Comment (mark-prins) [inline server/service/src/sync/test/integration/errors.rs:55]: required?

**Review by mark-prins — APPROVED** [review]
🎉  The error handling is very slick! Like what you've done here

Have a concern about the formatting.. when testing I'm not seeing the errors formatted as nicely as I thought they should be from reading the code. The information is there, it's just a bit hard to parse.

Also - if the password is wrong we don't get much help from the error:

```
2022-10-03 14:19:09.560466 ERROR service::sync::sync_status::logger - Error in sync: Error while pushing remote records  service/src/sync/sync_status/logger.rs:147
```

**Comment by andreievg** [comment]
@mark-prins for both of the printed error example, definitely should be better formatted, sorry I missed the cli output, was only testing what front end was displaying:

fixed: 460a0bdba7ce4c5cc57898bac8cfac70d12142c3
Screen Shot 2022-10-03 at 3 56 06 PM

Screen Shot 2022-10-03 at 2 49 40 PM

File: server/service/src/sync/api/core.rs [hunk 14]
```diff
@@ -122,9 +144,9 @@ impl SyncApiV5 {
 
 #[derive(Error, Debug)]
 pub enum ParsingResponseError {
-    #[error("Cannot retreive response body: {0}")]
-    CannotGetTextReponse(reqwest::Error),
-    #[error("Could not parse response body, error: ({source}) reponse; ({response_text}) ")]
+    #[error("Cannot retreive response body")]
+    CannotGetTextReponse(#[from] reqwest::Error),
```
[the remark this answers is not in the document]
Comment (andreievg) [reply server/service/src/sync/api/core.rs:148]: 🙏 did this straight into head pr: 6cc02a06f6f37b5417cc36a3a9d11e50ca19d113

File: server/service/src/sync/remote_data_synchroniser.rs (hunk 12 above)
Comment (andreievg) [reply server/service/src/sync/remote_data_synchroniser.rs:158]: Thanks, did this straight into head PR: bb8309272975b8f867a4d8f8c4cd2d87c73d93f2 (also changed the enum definition)

File: server/service/src/sync/test/integration/errors.rs (hunk 13 above)
Comment (andreievg) [reply server/service/src/sync/test/integration/errors.rs:55]: remove in head pr: bb8309272975b8f867a4d8f8c4cd2d87c73d93f2

File: server/service/src/sync/api/error.rs (hunk 9 above)
Comment (andreievg) [reply server/service/src/sync/api/error.rs:11]: This is definitely not what I intended as output, please see comment below: https://github.com/openmsupply/open-msupply/pull/671#issuecomment-1264868340. 🙏 for catching this
